File: mobile/modules/bluetooth-sdk/ios/Source/sgcs/Simulated.swift
//
//  Simulated.swift
//  AOS
//
//  Created by Matthew Fosse on 10/7/25.
//

@MainActor
class Simulated: SGCManager {
    init() {
        DeviceStore.shared.apply("glasses", "fullyBooted", true)
        DeviceStore.shared.apply("glasses", "connected", true)
        DeviceStore.shared.apply("glasses", "connectionState", ConnTypes.CONNECTED)
        DeviceStore.shared.apply("glasses", "micEnabled", false)
        DeviceStore.shared.apply(
            "glasses",
            "voiceActivityDetectionEnabled",
            BluetoothSdkDefaults.voiceActivityDetectionEnabled
        )
        DeviceStore.shared.apply("glasses", "bluetoothClassicConnected", false)
    }

    // MARK: - Device Information

    var type: String = DeviceTypes.SIMULATED
    var ready: Bool = true
    var connectionState: String = ConnTypes.CONNECTED

    var appVersion: String = ""
    var buildNumber: String = ""
    var deviceModel: String = ""
    var androidVersion: String = ""
    var otaVersionUrl: String = ""
    var firmwareVersion: String = ""
    var bluetoothMacAddress: String = ""
    var serialNumber: String = ""
    var style: String = ""
    var color: String = ""

    // MARK: - Hardware Status

    var hasMic: Bool = false
    var batteryLevel: Int = 100
    var headUp: Bool = false
    var micEnabled: Bool = false

    // MARK: - Case Status

    var caseOpen: Bool = false
    var caseRemoved: Bool = false
    var caseCharging: Bool = false
    var caseBatteryLevel: Int = -1

    // MARK: - Network Status

    var wifiSsid: String = ""
    var wifiConnected: Bool = false
    var wifiLocalIp: String = ""
    var hotspotEnabled: Bool = false
    var hotspotSsid: String = ""
    var hotspotPassword: String = ""
    var hotspotGatewayIp: String = ""

    // MARK: - Audio Control

    func setMicEnabled(_: Bool) {
        Bridge.log("setMicEnabled")
    }

    func sortMicRanking(list: [String]) -> [String] {
        return list
    }

    // MARK: - Messaging

    func sendJson(_: [String: Any], wakeUp _: Bool, requireAck _: Bool) {
        Bridge.log("sendJson")
    }

    // MARK: - Camera & Media

    func requestPhoto(_ request: PhotoRequest) {
        Bridge.log("requestPhoto save=\(request.save) sound=\(request.sound)")
    }

    func startStream(_: [String: Any]) {
        Bridge.log("startStream")
    }

    func stopStream() {
        Bridge.log("stopStream")
    }

    func sendStreamKeepAlive(_: [String: Any]) {
        Bridge.log("sendStreamKeepAlive")
    }

    func startVideoRecording(requestId _: String, save _: Bool, sound _: Bool) {
        Bridge.log("startVideoRecording")
    }

    func stopVideoRecording(requestId _: String) {
        Bridge.log("stopVideoRecording")
    }

    // MARK: - Button Settings

    func sendButtonPhotoSettings() {
        Bridge.log("sendButtonPhotoSettings")
    }

    func sendButtonVideoRecordingSettings() {
        Bridge.log("sendButtonVideoRecordingSettings")
    }

    func sendCameraFovSetting() {
        Bridge.log("sendCameraFovSetting")
    }

    func sendButtonMaxRecordingTime() {}

    // MARK: - Display Control

    func setBrightness(_: Int, autoMode _: Bool) {
        Bridge.log("setBrightness")
    }

    func clearDisplay() {
        Bridge.log("clearDisplay")
    }

    func sendText(_ text: String) async {
        await sendTextWall(text)
    }

    func sendTextWall(_: String) async {
        Bridge.log("sendTextWall")
    }

    func sendDoubleTextWall(_: String, _: String) async {
        Bridge.log("sendDoubleTextWall")
    }

    func displayBitmap(base64ImageData _: String, x _: Int32?, y _: Int32?, width _: Int32?, height _: Int32?) async -> Bool {
        Bridge.log("displayBitmap")
        return false
    }

    func showDashboard() {
        Bridge.log("showDashboard")
    }

    func setDashboardPosition(_: Int, _: Int) {
        Bridge.log("setDashboardPosition")
    }

    // MARK: - Device Control

    func setHeadUpAngle(_: Int) {
        Bridge.log("setHeadUpAngle")
    }

    func getBatteryStatus() {
        Bridge.log("getBatteryStatus")
    }

    func setSilentMode(_: Bool) {
        Bridge.log("setSilentMode")
    }

    func exit() {
        Bridge.log("exit")
    }

    func sendShutdown() {
        Bridge.log("sendShutdown - not supported on Simulated")
    }

    func sendReboot() {
        Bridge.log("sendReboot - not supported on Simulated")
    }

    func sendRgbLedControl(requestId: String, packageName _: String?, action _: String, color _: String?, onDurationMs _: Int, offDurationMs _: Int, count _: Int) {
        Bridge.log("sendRgbLedControl - not supported on Simulated")
        Bridge.sendRgbLedControlResponse(requestId: requestId, success: false, error: "device_not_supported")
    }

    // MARK: - Connection Management

    func disconnect() {
        Bridge.log("disconnect")
    }

    func forget() {
        Bridge.log("forget")
    }

    func findCompatibleDevices() {
        Bridge.log("findCompatibleDevices")
    }

    func stopScan() {
        Bridge.log("stopScan")
    }

    func connectById(_: String) {
        Bridge.log("connectById")
    }

    func getConnectedBluetoothName() -> String? {
        return nil
    }

    func cleanup() {
        Bridge.log("cleanup")
    }

    func ping() {
        Bridge.log("ping")
    }

    func dbg1() {}
    func dbg2() {}
    func connectController() {}
    func disconnectController() {}

    // MARK: - Network Management

    func requestWifiScan() {
        Bridge.log("requestWifiScan")
    }

    func sendWifiCredentials(_: String, _: String) {
        Bridge.log("sendWifiCredentials")
    }

    func forgetWifiNetwork(_ ssid: String) {
        Bridge.log("forgetWifiNetwork: \(ssid)")
    }

    func sendHotspotState(_: Bool) {
        Bridge.log("sendHotspotState")
    }

    func sendUserEmailToGlasses(_ email: String) {
        Bridge.log("sendUserEmailToGlasses: \(email)")
    }

    func sendOtaStart(otaVersionUrl: String?) {
        Bridge.log("sendOtaStart")
    }

    func sendOtaQueryStatus() {
        Bridge.log("sendOtaQueryStatus")
    }

    // MARK: - Gallery

    func queryGalleryStatus() {
        Bridge.log("queryGalleryStatus")
    }

    func sendGalleryMode() {
        Bridge.log("sendGalleryMode")
    }

    // MARK: - Version Info

    func requestVersionInfo() {
        Bridge.log("requestVersionInfo - not supported on Simulated")
    }

    func sendIncidentId(_: String, apiBaseUrl _: String?) {}
}

File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/sgcs/Simulated.kt
package com.mentra.bluetoothsdk.sgcs

import com.mentra.bluetoothsdk.BluetoothSdkDefaults
import com.mentra.bluetoothsdk.Bridge
import com.mentra.bluetoothsdk.DeviceManager
import com.mentra.bluetoothsdk.PhotoRequest
import com.mentra.bluetoothsdk.DeviceStore
import com.mentra.bluetoothsdk.utils.ConnTypes
import com.mentra.bluetoothsdk.utils.DeviceTypes

class Simulated : SGCManager() {

    init {
        type = DeviceTypes.SIMULATED
        DeviceStore.apply("glasses", "fullyBooted", true)
        DeviceStore.apply("glasses", "connected", true)
        DeviceStore.apply("glasses", "connectionState", ConnTypes.CONNECTED)
        DeviceStore.apply("glasses", "micEnabled", false)
        DeviceStore.apply(
            "glasses",
            "voiceActivityDetectionEnabled",
            BluetoothSdkDefaults.VOICE_ACTIVITY_DETECTION_ENABLED
        )
    }

    // Audio Control
    override fun setMicEnabled(enabled: Boolean) {
        Bridge.log("setMicEnabled")
    }

    override fun sortMicRanking(list: MutableList<String>): MutableList<String> {
        return list
    }

    // Camera & Media
    override fun requestPhoto(request: PhotoRequest) {
        Bridge.log("requestPhoto save=${request.save}, sound=${request.sound}")
    }

    override fun startStream(message: MutableMap<String, Any>) {
        Bridge.log("startStream")
    }

    override fun stopStream() {
        Bridge.log("stopStream")
    }

    override fun sendStreamKeepAlive(message: MutableMap<String, Any>) {
        Bridge.log("sendStreamKeepAlive")
    }

    override fun startVideoRecording(requestId: String, save: Boolean, sound: Boolean) {
        Bridge.log("startVideoRecording sound=$sound")
    }

    override fun stopVideoRecording(requestId: String) {
        Bridge.log("stopVideoRecording")
    }

    // Button Settings
    override fun sendButtonPhotoSettings() {
        Bridge.log("sendButtonPhotoSettings")
    }

    override fun sendButtonVideoRecordingSettings() {
        Bridge.log("sendButtonVideoRecordingSettings")
    }

    override fun sendButtonMaxRecordingTime() {
        Bridge.log("sendButtonMaxRecordingTime")
    }

    override fun sendCameraFovSetting() {
        Bridge.log("sendCameraFovSetting")
    }

    // Display Control
    override fun setBrightness(level: Int, autoMode: Boolean) {
        Bridge.log("setBrightness")
    }

    override fun clearDisplay() {
        Bridge.log("clearDisplay")
    }
    
    override fun sendText(text: String) {
        Bridge.log("sendText")
    }

    override fun sendTextWall(text: String) {
        Bridge.log("sendTextWall")
    }

    override fun sendDoubleTextWall(top: String, bottom: String) {
        Bridge.log("sendDoubleTextWall")
    }

    override fun displayBitmap(
            base64ImageData: String,
            x: Int?,
            y: Int?,
            width: Int?,
            height: Int?
    ): Boolean {
        Bridge.log("displayBitmap")
        return false
    }

    override fun showDashboard() {
        Bridge.log("showDashboard")
    }

    override fun setDashboardPosition(height: Int, depth: Int) {
        Bridge.log("setDashboardPosition")
    }

    // Device Control
    override fun setHeadUpAngle(angle: Int) {
        Bridge.log("setHeadUpAngle")
    }

    override fun getBatteryStatus() {
        Bridge.log("getBatteryStatus")
    }

    override fun setSilentMode(enabled: Boolean) {
        Bridge.log("setSilentMode")
    }

    override fun exit() {
        Bridge.log("exit")
    }

    override fun sendShutdown() {
        Bridge.log("sendShutdown - not supported on Simulated")
    }

    override fun sendReboot() {
        Bridge.log("sendReboot - not supported on Simulated")
    }

    override fun sendRgbLedControl(
            requestId: String,
            packageName: String?,
            action: String,
            color: String?,
            onDurationMs: Int,
            offDurationMs: Int,
            count: Int
    ) {
        Bridge.log("sendRgbLedControl - not supported on Simulated")
        Bridge.sendRgbLedControlResponse(requestId, false, "device_not_supported")
    }

    // Connection Management
    override fun disconnect() {
        Bridge.log("disconnect")
    }

    override fun forget() {
        Bridge.log("forget")
    }

    override fun findCompatibleDevices() {
        Bridge.log("findCompatibleDevices")
    }

    override fun stopScan() {
        Bridge.log("stopScan")
    }

    override fun connectById(id: String) {
    }

    override fun getConnectedBluetoothName(): String {
        return ""
    }

    override fun cleanup() {
        Bridge.log("cleanup")
    }

    override fun ping() {
        Bridge.log("ping")
    }

    override fun dbg1() {}
    override fun dbg2() {}

    // Network Management
    override fun requestWifiScan(scanId: String?) {
        Bridge.log("requestWifiScan")
    }

    override fun sendWifiCredentials(ssid: String, password: String) {
        Bridge.log("sendWifiCredentials")
    }

    override fun forgetWifiNetwork(ssid: String) {
        Bridge.log("forgetWifiNetwork: $ssid")
    }

    override fun sendHotspotState(enabled: Boolean) {
        Bridge.log("sendHotspotState")
    }

    override fun sendUserEmailToGlasses(email: String) {
        Bridge.log("sendUserEmailToGlasses: $email")
    }

    override fun sendIncidentId(incidentId: String, apiBaseUrl: String?) {
        Bridge.log("sendIncidentId: $incidentId")
    }

    // Gallery
    override fun queryGalleryStatus() {
        Bridge.log("queryGalleryStatus")
    }

    override fun sendGalleryMode() {
        Bridge.log("SIMULATED: 📸 Received gallery mode")
    }

    // Version info
    override fun requestVersionInfo() {
        Bridge.log("SIMULATED: 📱 Requesting version info (no-op)")
    }
}
